contracts/libraries/GnosisAuction.sol

```diff
@@ -124,47 +125,55 @@ library GnosisAuction {
 
     function placeBid(BidDetails memory bidDetails)
         internal
-        returns (uint256 oTokensToBuy, uint256 bidAmount)
+        returns (
+            uint96 sellAmount,
+            uint96 buyAmount,
+            uint64 userId
+        )
     {
         // calculate how much to allocate
-        bidAmount = bidDetails
-            .lockedBalance
-            .mul(bidDetails.optionAllocationPct)
-            .div(10000);
+        sellAmount = uint96(
+            bidDetails.lockedBalance.mul(bidDetails.optionAllocationPct).div(
+                10000
+            )
+        );
 
-        // divide the `asset` bidAmount by the target premium per oToken to
+        // divide the `asset` sellAmount by the target premium per oToken to
         // get the number of oTokens to buy (8 decimals)
-        oTokensToBuy = bidAmount.div(bidDetails.optionPremium).mul(10**8).div(
-            10**bidDetails.assetDecimals
+        buyAmount = uint96(
+            uint256(sellAmount)
+                .mul(10**bidDetails.assetDecimals)
+                .div(bidDetails.optionPremium)
+                .mul(10**8)
+                .div(10**bidDetails.assetDecimals)
         );
 
         require(
-            bidAmount <= type(uint96).max,
-            "bidAmount > type(uint96) max value!"
+            sellAmount <= type(uint96).max,
+            "sellAmount > type(uint96) max value!"
         );
         require(
-            oTokensToBuy <= type(uint96).max,
-            "oTokensToBuy > type(uint96) max value!"
+            buyAmount <= type(uint96).max,
+            "buyAmount > type(uint96) max value!"
```

Comment:
Checking for uint96 overflow here doesn't actually do anything. We need to do the checks before casting it to uint96.

This works:

```solidity
uint256 number = someNumber;
require(number <= type(uint96.max));
uint96 smallNumber = uint96(number);
```

This does not work:

```solidity
uint256 number = someNumber;
// casting already truncates the integer to be the size uint96.
uint96 smallNumber = uint96(number); 
require(smallNumber <= type(uint96.max));
```

Reply:
yes good point